articles/WideStrings.md

```diff
@@ -0,0 +1,264 @@
+of bytes.  The second way it could be mapped is onto the "wchar/wstring" type
+defined by the DDS specification.  The DDS specification itself doesn't define
+a size for wchar, but the X-Types specification (for DDS introspection) does
+say that a wchar should be mapped onto a Char32.  Thus, all known encodings
+would fit into a wchar, and DDS introspection tools could look into wstrings.
+Thus, the current design of wide strings map onto wchar DDS types.
+
+**Runtime impact of wide string**
+
+Dealing with wide strings puts more strain on the software of a system, both in
+terms of speed and in terms of code size.  This is felt most acutely when talking
+about small microcontrollers, where both flash size (code space) and performance
+(processor speed) are at a premium.  Some of the common encodings (including UTF-8)
+are defined as being variable width, meaning that a character can take 1, 2, or 4
+characters to represent.  These encoding also typically have the ability to combine
+one character with the next in non-trivial ways.  As one of the goals of ROS 2.0
+is to support small, constrained systems, dealing with wide strings may be out of
+the question.  Thus, the current ROS 2.0 design for wide strings makes them optional,
+defines them as a separate type from regular strings, and recommends against
+using them for common messages.
+
+### What is the encoding of the wide string?
+
+There are two main ways that ROS 2.0 can provide for the encoding of a wide string.  Either
+ROS 2.0 can define exactly what the encoding will be for all strings, or it can allow the
+user to specify the encoding when handing it to ROS 2.0.  There are pros and cons to each
+approach.
+
+**ROS 2.0 defines the encoding**
+
+Pros:
+
+* All strings known to be in the same encoding.
+
+Cons:
+
+* If the user-level code uses a different encoding, the user is responsible
+for doing a conversion to the ROS 2.0 defined encoding.  This can involve
+some runtime cost.
+
+**User defines the encoding**
+
+Pros:
+
+* No conversions necessary for user-to-ROS 2 API.
+
+Cons:
+
+* The user must always tell ROS 2.0 what encoding the data is in.
+* The encoding needs to be transmitted on the wire to the other side.
+* It is not clear how ROS 2.0 will transmit the encoding.  Fixed list?  Which
+encodings go into the list?
+
+While a user defined encoding looks somewhat attractive at first, the downsides
+to it make it difficult to determine exactly which encodings are in use.  The
+downsides also mean that any program that wants to parse ROS 2 messages (or bag
+files) may potentially have to deal with many different encodings.  Additionally,
+Unicode (and in particular, UTF-8) are very commonplace now, while other encodings
+are becoming more esoteric.  Even earlier Unicode (UTF-16 and UTF-32) encodings are
+essentially deprecated in favor of UTF-8.  For these reasons, the current design of
+the ROS 2.0 wide string is to use a UTF-8 encoding for all strings.
```

Comment:
It would be good to address these separately. It's quite possible our "wide string" datatype could implicitly do these conversions automatically.